### ISSUE TITLE
Add more specific IDs to QuickFix Code Actions for keybindings

### ISSUE THREAD

**Issue by DanTup** [Dart-Code/Dart-Code#561]

**Comment by DanTup** [comment]
@mjbvz we're trying to add IDs in for quick fixes (and later refactors) for Dart. It seems like there aren't that many in use for TS/JS so we don't have much to copy (we'd like to try and keep things aligned with other extensions to allow existing keybindings the user has set up to work, but we'd also like to haev specific enough IDs that if a user wants to bind a specific action they can).

You can see the full list of assists we support (refactors aren't done yet) here:

https://github.com/dart-lang/sdk/blob/master/pkg/analysis_server/lib/src/services/correction/assist.dart#L33

Do you have any suggestions on the naming/categories? I guess we should drop `dart.` from the front (if not in the language server, in the provider) but otherwise do they seem sane (eg. the idea of calling these things "assist" and the groupings like "convert", "wrap", "surround"?)

**Comment by DanTup** [comment]
This seems to work, eg. if I had this to my keybindings it'll invoke:

```js
{
	"key": "ctrl+oem_comma",
	"command": "editor.action.codeAction",
	"args": {
		"kind": "refactor.flutter.wrap.padding"
	}
}
```

We need to document this (with a full list of the code action kinds).

**Comment by bwilkerson** [comment]
Do we have good id's (that we're happy to live with)?

Do we want the documentation in the DartCode docs, or in the Dart tooling docs somewhere (that the DartCode docs can link to)? My inclination would be the latter, because it will probably be easier to maintain it as we add new codes.

**Comment by DanTup** [comment]
@bwilkerson I think we were happy with the IDs. It'd be great to not have to maintain them on my though the prefix in Dart Code is changed (to `refactor.`). I was going to dump the list in the readme for now (so I can ship this) prefixed with a note they they may be updated over time, and if we get them somewhere centrally I can always link to it with an explanation about the IDs?

**Comment by DanTup** [comment]
@bwilkerson This is what I was going to do for now -> https://github.com/Dart-Code/Dart-Code#refactoring-kinds

LMK what you think

**Comment by bwilkerson** [comment]
That looks good to me. I'm not sure we want to change the prefix, but that's your call. (Leaving it as is would require less explanation from you and possibly less confusion for your users.)

I'm guessing that we're not going to add ids for fixes, given that those only apply when there's an error. Are we going to add ids for what server uses 'refactoring' for? (If so, there's also a risk that the prefix change could lead to a conflict.)

**Comment by DanTup** [comment]
> That looks good to me. I'm not sure we want to change the prefix, but that's your call

I really wanted to try and use the Code convention so there could be overlap with other languages. For example the usual keybind is `Ctrl+.` but it does fixes and refactors. If the user wanted to separate them into two different keybinds (so they have easy access to refactors, say) then by using the Code prefixes it'll work across different languages. If we stick with our own, they've have to bind the same key to two prefixes (I don't know if that's possible).

> I'm guessing that we're not going to add ids for fixes, given that those only apply when there's an error

I was going to (but haven't yet) just set them all to [`CodeActionKind.QuickFix`](https://code.visualstudio.com/docs/extensionAPI/vscode-api#CodeActionKind) for the reason above - so the user can at least create a keybind for quickfixes (without refactors).

> Are we going to add ids for what server uses 'refactoring' for? (If so, there's also a risk that the prefix change could lead to a conflict.)

We haven't implemented them yet, but when we do I think we should. I'm not sure if we'd end up with overlap (if we did, it suggests we have two things doing the same job?) but if we did that wouldn't be a problem, the user would just see a menu with the two items in (assuming they were both valid at the same place in code - if only one is, by default it'll just execute without showing a menu - though the use can control this).
